### ISSUE TITLE
spdlog and fmt staticness

### ISSUE THREAD

**Issue by lukester1975** [mesonbuild/wrapdb#881]
I'm not sure if this is expected behaviour, but if you configure this https://github.com/lukester1975/spdlog-fmt on a system without spdlog and fmt (such that the wraps are used) you end up with spdlog as a static library and fmt as dynamic due to spdlog's fmt dep.

It's trivially fixable by swapping the order, but that seems a little fragile. Or by setting a project-wide `default_library=static`, but that's a little draconian for my taste.

I don't know if this would be considered a suitable approach to tidy this up (if it's even considered an issue), but this worked: https://github.com/lukester1975/wrapdb/commit/f0ed5b02a04bbdba02232203f6f44cba1575b094

Thanks

**Comment by metab0t** [comment]
A simpler example:
```meson
project('spdlog-fmt', 'cpp', version: '0.0.0')

fmt1 = dependency('fmt')
fmt2 = dependency('fmt', static: true)

executable('spdlog-fmt', 'main.cpp', dependencies: [fmt2])
```
The order of `fmt1` and `fmt2` matters, and it seems that only the first one is selected.

**Comment by xclaesse** [comment]
yes, order matters, more generally it is the responsability of the main project to ensure (sub-)subprojects are configured with the options it wants. Similar thing happens if you do:
```meson
dependency('foo', default_options: 'opt=true')
dependency('foo', default_options: 'opt=false')
```

**Comment by lukester1975** [comment]
As an end-user, I was a little surprised that dependency-specified `static`ness doesn't propagate down to its deps without having to make it do so. I don't think I'd ever want to build a static library that itself ends up building a shared one. My head would explode! I guess there's reasons it is the way it is, but seems saner if it did propagate.

I would also not mind erroring out during configure if there are clashes. My example was trivial but if there's some distance between clashing deps (especially if not obvious the same dependency has appeared twice, due to sub deps) then it could be non-trivial to figure out. More head related explosions. Or an appropriate warning?

Thanks

**Comment by metab0t** [comment]
https://github.com/mesonbuild/meson/issues/11315#issuecomment-1398539249

TLDR: Set `fmt:default_library=static` to ensure it is built as a static library.

**Comment by xclaesse** [comment]
> As an end-user, I was a little surprised that dependency-specified `static`ness doesn't propagate down to its deps

I understand the confusion. The problem is inheriting options from the "parent" subproject does not eliminate the ordering problem, you can have A->B->C and A->D->C, and you get back to initial problem, does C get options from B or D? It depends on which order A lookup dependencies B and D. Meson gets away from that problem by leaving it completely up to the main project (default_options: 'C:default_library=static') or user (-DC:default_library=static).

**Comment by lukester1975** [comment]
Sure, if the options clashed. But when they don't... It also seems strange to have to know about transitive deps to get it to work sensibly (ok, spdlog here is a bad example since it is explicitly asking for fmt, but in general).

Anyway, it's fine. Just mentioning it in case it was not intentional or not considered. I'm happy with any work around and waiting for conan v2 to mature a bit, that seems to have pretty good meson integration!

Thanks

**Comment by eli-schwartz** [comment]
> It also seems strange to have to know about transitive deps to get it to work sensibly

What you probably want here is an option to make all subprojects (but not the main project) default to building with default_library set to static. :)

But that is still very much a user choice, and people want different things.

(-Dprefer_static=true exists, now, but isn't subproject specific.)

**Comment by lukester1975** [comment]
Oh I'm never against configurability, unless I turn that option off.

I'm just saying if `static` on a dep propagated to all its child deps, then to me that would seem to be a saner default. If someone wanted something different then the fact that those options **can** be set per-dep is still there...

Anyhoo I might as well close this.

Cheers
